Improved the agent demand calculation logic.

tamizhgeek opened the pull request:
A new agent is launched only when all of the below are true:
1. The scheduled pending jobs should be either positive or constantly greater than 1, for last 5 times.
2. The idle agents in the server should be 0.

@sriram-r @ashwanthkumar

tamizhgeek committed:
Improved the agent demand calculation logic.

A new agent is launched only when all of the above are true:
1. The scheduled pending jobs should be either positive or constantly greater than 1, for last 5 times.
2. The idle agents in the server should be 0.

tamizhgeek committed:
Separate config to enable/disable authentication with the GoCD API

ashwanthkumar commented on src/main/scala/com/indix/mesos/GOCDPoller.scala:
```diff
@@ -28,19 +52,48 @@ case class GOCDPoller(conf: FrameworkConfig) {
   def goIdleAgentsCount() = {
+    println("Polling Go server for idle agents")
+    try {
+      val responseJson = jsonRequest(buildUrl() + "/go/api/agents")
```
@tamizhgeek I know the api.go.cd response says this is the schema of the response, but can you quickly check the same on our servers? Our main build is running the latest 16.2.0. They look different actually. We might want to file a bug against it.

tamizhgeek replied:
@ashwanthkumar The difference in the responses are because of the Accept header we send. The difference is this:

`curl -X GET http://build.gocd.io/go/api/agents -H "Authorization: Basic asdadsadasdasd="`

and 

`curl -X GET http://build.gocd.io/go/api/agents -H "Authorization: Basic adsdasdasda=" -H "Accept: application/vnd.go.cd.v1+json"`

:)

ashwanthkumar commented:
Some questions (pardon my ignorance)
- From the set of changes, we don't seem to be doing Environment(s) and Resource(s) based scaling up. 
- Also we don't seem to be downscaling the agents.
Are these features already present?

ashwanthkumar commented on src/main/scala/com/indix/mesos/GOCDPoller.scala:
```diff
@@ -28,19 +52,48 @@ case class GOCDPoller(conf: FrameworkConfig) {
       responseHistory += scheduled
-    
+    }
     if(responseHistory.size > 5) {
```
I would really like the `5` number (scheduledJobs > 5) to be a configuration :)

ashwanthkumar commented on src/main/scala/com/indix/mesos/GOCDPoller.scala:
```diff
@@ -28,19 +52,48 @@ case class GOCDPoller(conf: FrameworkConfig) {
+        responseHistory.clear()
+      } else {
+        println("Idle agents found. Not launching any new Go Agent now.")
+        responseHistory.drop(0)
```
Why do we do a `drop(0)` here?

tamizhgeek replied:
My idea is to always keep the last 5(basically, any n) responses in the history before taking decisions. Once the decision is taken (upscale/downscale), we start from a clean slate. Would love to have other ideas on how to implement it better.

brewkode replied:
Azhagu - looking at last 5 polls for job queue makes sense. But, starting
clean after a decision is made, doesn't seem to. Shouldn't you dropping the
oldest one and still use the rest of them..?

On Sat, Feb 20, 2016 at 8:41 PM, Azhaguselvan SP notifications@github.com
wrote:

> In src/main/scala/com/indix/mesos/GOCDPoller.scala
> https://github.com/ind9/gocd-mesos/pull/3#discussion_r53551737:
> 
> > ```
> >  if(responseHistory.size > 5) {
> > ```
> > -      println(s"More than 5 jobs pending in the GOCD. queuing a new agent launch now.")
> > -      TaskQueue.enqueue(GoTask("", conf.goAgentDocker, ""))
> > -      responseHistory.clear()
> > -      println(s"More than 5 jobs pending in the GOCD. checking if any agents are idle.")
> > -      if(goIdleAgentsCount() == 0 && isDemandPositive()) {
> > -        println("No idle agents found. Launching a new agent launch now.")
> > -        TaskQueue.enqueue(GoTask("", conf.goAgentDocker, ""))
> > -        responseHistory.clear()
> > -      } else {
> > -        println("Idle agents found. Not launching any new Go Agent now.")
> > -        responseHistory.drop(0)
> 
> My idea is to always keep the last 5(basically, any n) responses in the
> history before taking decisions. Once the decision is taken
> (upscale/downscale), we start from a clean slate. Would love to have other
> ideas on how to implement it better.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ind9/gocd-mesos/pull/3/files#r53551737.

tamizhgeek replied:
@sriram-r  Hmmm. Once a new agent is launched, we don't expect the jobs to get scheduled immediately. There is some lag between that. If we don't clean the history, isn't there is a possibility that before this agent is launched and the job is scheduled, another one might also get launched?

brewkode replied:
Generally, that's handled via "UPSCALING" state in your GoCD mesos
framework.

On Sat, Feb 20, 2016 at 8:54 PM, Azhaguselvan SP notifications@github.com
wrote:

> In src/main/scala/com/indix/mesos/GOCDPoller.scala
> https://github.com/ind9/gocd-mesos/pull/3#discussion_r53551958:
> 
> > ```
> >  if(responseHistory.size > 5) {
> > ```
> > -      println(s"More than 5 jobs pending in the GOCD. queuing a new agent launch now.")
> > -      TaskQueue.enqueue(GoTask("", conf.goAgentDocker, ""))
> > -      responseHistory.clear()
> > -      println(s"More than 5 jobs pending in the GOCD. checking if any agents are idle.")
> > -      if(goIdleAgentsCount() == 0 && isDemandPositive()) {
> > -        println("No idle agents found. Launching a new agent launch now.")
> > -        TaskQueue.enqueue(GoTask("", conf.goAgentDocker, ""))
> > -        responseHistory.clear()
> > -      } else {
> > -        println("Idle agents found. Not launching any new Go Agent now.")
> > -        responseHistory.drop(0)
> 
> @sriram-r https://github.com/sriram-r Hmmm. Once a new agent is
> launched, we don't expect the jobs to get scheduled immediately. There is
> some lag between that. If we don't clean the history, isn't there is a
> possibility that before this agent is launched and the job is scheduled,
> another one might also get launched?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ind9/gocd-mesos/pull/3/files#r53551958.

tamizhgeek replied:
So, during "UPSCALING" state, we don't schedule any more agent launches. Thanks @sriram-r will think in those lines :+1:

brewkode replied:
So, this is what I had in my mind:
- Poll for job queue status
- If there's an upscale action needed, determine how many agents need to be
  brought up.
- Issue scale event for the # of agents
- Mark the Scalar state to UPSCALING
- Meanwhile, polling continues, but, use it only to update history, not to
  trigger events.
- Mark Scalar state to IDLE once upscaling completes
- Everything continues to chug along

On Sat, Feb 20, 2016 at 8:56 PM, Sriram Ramachandrasekaran <sriram@indix.com

> wrote:
> 
> Generally, that's handled via "UPSCALING" state in your GoCD mesos
> framework.
> 
> On Sat, Feb 20, 2016 at 8:54 PM, Azhaguselvan SP <notifications@github.com
> 
> > wrote:
> > 
> > In src/main/scala/com/indix/mesos/GOCDPoller.scala
> > https://github.com/ind9/gocd-mesos/pull/3#discussion_r53551958:
> > 
> > > ```
> > >  if(responseHistory.size > 5) {
> > > ```
> > > -      println(s"More than 5 jobs pending in the GOCD. queuing a new agent launch now.")
> > > -      TaskQueue.enqueue(GoTask("", conf.goAgentDocker, ""))
> > > -      responseHistory.clear()
> > > -      println(s"More than 5 jobs pending in the GOCD. checking if any agents are idle.")
> > > -      if(goIdleAgentsCount() == 0 && isDemandPositive()) {
> > > -        println("No idle agents found. Launching a new agent launch now.")
> > > -        TaskQueue.enqueue(GoTask("", conf.goAgentDocker, ""))
> > > -        responseHistory.clear()
> > > -      } else {
> > > -        println("Idle agents found. Not launching any new Go Agent now.")
> > > -        responseHistory.drop(0)
> > 
> > @sriram-r https://github.com/sriram-r Hmmm. Once a new agent is
> > launched, we don't expect the jobs to get scheduled immediately. There is
> > some lag between that. If we don't clean the history, isn't there is a
> > possibility that before this agent is launched and the job is scheduled,
> > another one might also get launched?
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/ind9/gocd-mesos/pull/3/files#r53551958.

tamizhgeek commented:
@ashwanthkumar Yeah. We don't do any downscaling as of now. Also the agents are not launched specific to environment/resources. They are TODO stuff.

tamizhgeek committed:
Slightly better demand/supply management

tamizhgeek committed:
New Scalar implementation to orchestrate agent scaling up/down

1. The scalar runs in a thread instead of the poller and supports scale down and scaleup.
2. The agents are remembered using a unique UUID after launch. This helps to find the agents launched by us vs other systems.

tamizhgeek committed:
Added specs for the new Scalar implementation
